src/index.js

```diff
@@ -349,131 +278,246 @@ export default class RestClient {
     }
 
-    if (errors) return { errors, status: res.status };
-    let body = await res[options.streamReader]();
-    body = options.bodyParser(body, context);
-    if (body.errors) return { errors: body.errors, status: res.status };
-    return { data: body.data, headers: res.headers, status: res.status };
+    if (responseGroup === 'serverError') {
+      return this._handleRetry(
+        method, endpoint, fetchOptions, context, options, headers, _metadata,
+      );
+    }
+
+    if (!headers.get('content-type')) {
+      requests[fetchID].push(_metadata);
+      return { headers, metadata: _metadata };
+    }
+
+    const parsed = options.bodyParser(await res[options.streamReader](), context);
+
+    if (parsed.errors) {
+      errors = castArray(parsed.errors);
+      _metadata.errors = [..._metadata.errors, ...errors];
+    }
+
+    requests[fetchID].push(_metadata);
+    return { data: parsed.data, headers, metadata: _metadata };
   }
 
   /**
    *
    * @private
    * @param {string} endpoint
-   * @param {string|Array<string>} values
+   * @param {Array<string>} values
    * @param {Object} context
    * @param {Object} options
    * @param {string} resourceKey
    * @return {Promise}
    */
   async _get(endpoint, values, context, options, resourceKey) {
-    const res = await this._fetch('GET', endpoint, context, options);
-    const data = res.data ? castArray(res.data) : [];
-    const errors = res.errors;
-    const _values = values ? castArray(values) : [];
-    this._resolveRequests({ data, errors }, _values, context, resourceKey);
-    if (errors) return errors;
-
-    if (data.length) {
-      this._setCache(data, res.headers, context.path, context.queryParams, resourceKey);
+    const cache = await this._checkCache(endpoint);
+    const { cacheability } = cache;
+    let res;
+
+    if (!cacheability || cacheability.noCache || !cacheability.check()) {
+      const headers = { ...options.headers };
+      const etag = cacheability && cacheability.etag;
+      if (etag) headers['If-None-Match'] = etag;
+      const method = 'GET';
+      const fetchOptions = { headers: new Headers(headers), method };
+      const metadata = { errors: [], resource: values };
+      res = await this._fetch(method, endpoint, fetchOptions, context, options, metadata);
+    } else {
+      const { requests } = this._getTracker(context.path);
+      const fetchID = context.fetchID;
+      if (!requests[fetchID]) requests[fetchID] = [];
+      requests[fetchID].push({ endpoint, fromCache: true, resource: values });
     }
 
-    if (!context.resource) return data;
-    return this._resolveResource(data, resourceKey, context.resource.active);
+    const metadata = res ? res.metadata : {};
+    const data = !res || metadata.status === 304 ? cache.data : res.data;
+    if (metadata.ok || metadata.status === 304) this._setCacheEntry(endpoint, data, res.headers);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') TypeError can be thrown as "res" might be null or undefined here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2259)